DM: failed to mock downstream table by create table statement in schema tracker

### What did you do?

create table with shard_row_id_bits and primary int key(tidb_enable_clusterd_index: INT_ONLY) in downstream

### What did you expect to see?

dm runs well

### What did you see instead?

[2023/12/14 10:49:47.826 +08:00] [ERROR] [tracker.go:413] ["Init dowstream schema info error. "] [task=dglp_g1] [unit="binlog replication"] [tableID=`ti_dglpdb`.`TM_GL_TXN_ITEM`] [error="[code=44018:class=schema-tracker:scope=internal:level=high], Message: failed to mock downstream table by create table statement in schema tracker, RawCause: [ddl:8200]Unsupported shard_row_id_bits for table with primary key as row id\ngithub.com/pingcap/tiflow/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/tiflow/dm/pkg/terror/terror.go:324\ngithub.com/pingcap/tiflow/dm/pkg/schema.(*downstreamTracker).getTableInfoByCreateStmt\n\tgithub.com/pingcap/tiflow/dm/pkg/schema/tracker.go:469\ngithub.com/pingcap/tiflow/dm/pkg/schema.(*downstreamTracker).getOrInit\n\tgithub.com/pingcap/tiflow/dm/pkg/schema/tracker.go:411\ngithub.com/pingcap/tiflow/dm/pkg/schema.(*Tracker).GetDownStreamTableInfo\n\tgithub.com/pingcap/tiflow/dm/pkg/schema/tracker.go:383\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).genAndFilterInsertDMLs\n\tgithub.com/pingcap/tiflow/dm/syncer/dml.go:113\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).handleRowsEvent\n\tgithub.com/pingcap/tiflow/dm/syncer/syncer.go:2554\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).Run\n\tgithub.com/pingcap/tiflow/dm/syncer/syncer.go:2307\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).Process\n\tgithub.com/pingcap/tiflow/dm/syncer/syncer.go:734\ngithub.com/pingcap/tiflow/dm/syncer.(*Syncer).Resume\n\tgithub.com/pingcap/tiflow/dm/syncer/syncer.go:3197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]

### Versions of the cluster

DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):
```console
v6.5.3
```
Upstream MySQL/MariaDB server version:
```console
(paste upstream MySQL/MariaDB server version here)
```
Downstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):
```console
(paste TiDB cluster version here)
```
How did you deploy DM: tiup or manually?
```console
(leave TiUP or manually here)
```
Other interesting information (system version, hardware config, etc):
```console
>
>
```

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

--- a/dm/pkg/schema/tracker.go
+++ b/dm/pkg/schema/tracker.go
@@ -32,6 +32,7 @@ import (
 	"github.com/pingcap/tidb/parser/model"
 	"github.com/pingcap/tidb/parser/mysql"
 	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/filter"
 	"github.com/pingcap/tidb/util/mock"
@@ -463,6 +464,17 @@ func (dt *downstreamTracker) getTableInfoByCreateStmt(tctx *tcontext.Context, ta
 	if err != nil {
 		return nil, dmterror.ErrSchemaTrackerInvalidCreateTableStmt.Delegate(err, createStr)
 	}
+
+	// suppress ErrTooLongKey
+	strictSQLModeBackup := dt.se.GetSessionVars().StrictSQLMode
+	dt.se.GetSessionVars().StrictSQLMode = false
+	// support drop PK
+	enableClusteredIndexBackup := dt.se.GetSessionVars().EnableClusteredIndex
+	dt.se.GetSessionVars().EnableClusteredIndex = variable.ClusteredIndexDefModeOff
+	defer func() {
+		dt.se.GetSessionVars().StrictSQLMode = strictSQLModeBackup
+		dt.se.GetSessionVars().EnableClusteredIndex = enableClusteredIndexBackup
+	}()
 
 	ti, err := ddl.BuildTableInfoWithStmt(dt.se, stmtNode.(*ast.CreateTableStmt), mysql.DefaultCharset, "", nil)
 	if err != nil {

--- a/dm/pkg/schema/tracker_test.go
+++ b/dm/pkg/schema/tracker_test.go
@@ -613,7 +613,7 @@ func TestGetDownStreamIndexInfo(t *testing.T) {
 
 	mock.ExpectQuery("SHOW CREATE TABLE " + tableID).WillReturnRows(
 		sqlmock.NewRows([]string{"Table", "Create Table"}).
-			AddRow("test", "create table t(a int primary key, b int, c varchar(20000), key(c(20000)))"))
+			AddRow("test", "create table t(a int, b int, c varchar(20000), primary key(a, b), key(c(20000)))/*!90000 SHARD_ROW_ID_BITS=6 */"))
 	dti, err := tracker.GetDownStreamTableInfo(tcontext.Background(), tableID, oriTi)
 	require.NoError(t, err)
 	require.NotNil(t, dti.WhereHandle.UniqueNotNullIdx)

